### ISSUE TITLE
Analysis extremely slow

### ISSUE THREAD

**Issue by bex605** [DeepLabCut/DeepLabCut#3384]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating System

Windows 11 Pro

### DeepLabCut version

3.0.0

### What engine are you using?

pytorch

### DeepLabCut mode

single animal

### Device type

GPU

### Bug description 🐛

I am experiencing extremely slow video analysis in DeepLabCut 3.0.0.

When I click "Analyze Videos", analyzing a single video takes approximately 2 hours and 44 minutes for a video that is only 5 minutes long (about 6,000 frames at 20 fps). This seems much slower than I would expect.

While the analysis is running, I checked the Windows Task Manager and the GPU appears to be used correctly, showing approximately 70–100% utilization most of the time.

The rest of the workflow appears to work normally. I was able to:

- extract and label frames,
- train the network successfully (200 epochs took about 2 hours),
- extract outlier frames,
- relabel them, and
- retrain the network.

However, after retraining, running "Analyze Videos" is still extremely slow.

I also noticed that the tracking quality did not improve much after the outlier extraction and retraining. I am not sure whether this is related to the slow analysis or represents a separate issue, but I wanted to mention it.

I'm relatively new to DeepLabCut and not very experienced with Python or debugging, so I'd really appreciate explanations in simple terms if possible. Thank you!

### Steps To Reproduce

1. Create a new DeepLabCut 3.0.0 project.
2. Extract and label frames.
3. Train the network (training completes successfully).
4. Run Analyze Videos on a 5-minute video (approximately 6,000 frames at 20 fps).
5. Observe that the analysis takes about 2 hours and 44 minutes to complete.
6. Extract outlier frames from the same video, relabel them, and retrain the network.
7. Run Analyze Videos on the same video again and observe that the analysis remains similarly slow.

### Relevant log output

```shell

```

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by C-Achard** [comment]
Hello @bex605,

Thank you for the report. This is indeed slower than one would expect on GPU; just a few questions in order to determine whether this is a code or hardware problem:

- What kind of video? Resolution, codec, file format mainly. It could be the reads from video that are slow for this particular file.
-  What is the GPU model and CUDA version? `ipython` > `print("PyTorch CUDA version:", torch.version.cuda)`
- Can you run `nvidia-smi` while running the analysis and share the output? This would help us know how the memory is allocated by torch.
- Where are the files stored (SSD, network drive, ...)
- Which model?
- Does this also happen for other or shorter videos?
- Do you have any terminal output that may be relevant (warnings, even regular logs)

Finally, the performance is likely tied to the training data more than the analysis itself. It may be that your initial training run already yielded good results (hopefully!). Feel free to share some training logs and eval metrics if you'd like, happy to have a look.

Thanks again for the report and best,
Cyril

**Comment by bex605** [comment]
Hi Cyril,

thank you very much for your quick reply and for taking the time to look into this!

Here are the answers to your questions:

1. Video details
Resolution: 1920 × 1200
Frame rate: 20 fps
File format: .mp4
Codec: MPEG-4 Video (mp4v)

2. GPU model and CUDA version
GPU: NVIDIA GeForce GTX 1650
PyTorch CUDA version: 12.1

3. nvidia-smi output
I have attached a screenshot of the nvidia-smi output recorded while attempting to analyze a second video (the one that subsequently produced a Triton/Inductor error - see below). Unfortunately, I did not capture nvidia-smi during the analysis of the first video.
During the analysis of the first video, however, I monitored the GPU using the Windows Task Manager and observed GPU utilization of approximately 70–100% for most of the analysis.

4. File location
All project files and videos are stored locally on my computer (Desktop), not on a network drive.

5. Model
Engine: PyTorch.

6. Does this also happen for other or shorter videos?
With the first (and so far only) video that completed analysis, there were no error messages. The analysis finished successfully, but took approximately 2 hours and 44 minutes for a 5-minute video.
To answer your question, I then tried analyzing a second video with similar properties (same setup, similar duration, resolution, and frame rate). This time, however, the analysis did not progress beyond 1% (so far) and instead produced the attached error message. Therefore, I cannot yet determine whether the original slow-analysis issue also affects other videos, as the second analysis failed before completion.

7. Terminal output
I have attached the complete error output (ERROR_triton.txt) from the second analysis attempt.

Regarding the tracking performance: I understand that this is likely more related to the training than the analysis itself. Since my main concern at the moment is the unexpectedly slow analysis, I thought it would be best to first clarify whether there is an underlying software or hardware issue. Once that is resolved, I'd be happy to share the training logs and evaluation metrics as well if they would still be helpful.

Thanks again for your help!

Best,
Rebecca

<img width="861" height="466" alt="Image" src="https://github.com/user-attachments/assets/e2ff2683-b3b3-4aaa-9bdf-1efa3527dc7e" />

[ERROR_triton.txt](https://github.com/user-attachments/files/29628973/ERROR_triton.txt)

**Comment by C-Achard** [comment]
Okay, thank you for the info.

For the Triton compilation issue, can you save this as a .py and run it with python in your env?
```python
import torch._dynamo
torch._dynamo.config.suppress_errors = True

import deeplabcut

config = r"C:\Users\Rebecca\Desktop\RF\PhD\RawData\FFE_cycle_pylon\YOUR_PROJECT\config.yaml"
video = r"C:\Users\Rebecca\Desktop\RF\PhD\RawData\FFE_cycle_pylon\FFE_cycle_7848\FFE_cycle_7848_day1\Basler_acA1920-40um__24622089__20260529_151329184.mp4"

deeplabcut.analyze_videos(
    config,
    [video],
    videotype=".mp4",
    batch_size=1,
)
```

For the analysis, it could be that the batch size is too high for your GPU, which has relatively little memory. The script above sets a smaller number just to be sure.

Otherwise, I would still recommend:

1. Trying a 10–20 second clip from the same video.
2. Try a downsampled copy of the video, if that resolution is still sufficient for the bodyparts, and please share nvidia-smi during a working analysis run (if there are further errors let me know)
3. Could you also share the actual model architecture from the project’s pytorch_config.yaml, for example resnet_50, hrnet_w18, hrnet_w32, etc. rather than the engine?

Thanks!

**Comment by bex605** [comment]
Hi Cyril,

thanks again for your suggestions!

I ran the script you provided (torch._dynamo.config.suppress_errors = True, batch_size=1).

This time the analysis started successfully without the Triton/Inductor crash and used the trained model:

Model architecture: ResNet (resnet50_gn)

However, the analysis is still much slower than I would expect.

To test whether this was specific to the original video, I also created a 10–20 second clip from the same recording and analyzed it using the same script. The performance was very similar: it took approximately 19 seconds to process 10% of the clip, which corresponds to roughly 3 minutes for the entire 20-second clip.

While the analysis was running, I captured another nvidia-smi screenshot. It shows:

GPU utilization: 99%
GPU memory usage: 779 MiB / 4096 MiB

So it appears that the GPU is being utilized correctly, but the inference itself is still unexpectedly slow.

I've attached the updated nvidia-smi screenshot.

Please let me know if there is anything else I can test or any additional information that would be helpful.

Thanks again for your help!

Best,
Rebecca

<img width="834" height="441" alt="Image" src="https://github.com/user-attachments/assets/7b2d32c0-2cfb-4964-88b9-21f0b7c966b1" />

**Comment by deruyter92** [comment]
Hi @bex605, 

Actually, given your hardware, I think that full frame analysis with a bottom-up ResNet might just be inherently slow. What you could try to do to optimize:
- Increase the batch size a bit, `batch_size=1` was good for debugging but at least `batch_size=4` should maybe be doable with your hardware. 
- downscale the video resolution before analysis (1920×1200 might be a lot more than necessary, depending on your analysis)
- Try cropping, if the animal stays in a fixed region. (or consider [dynamic cropping](https://deeplabcut.github.io/DeepLabCut/docs/main-workflows/user-guide.html#i-analyze-new-videos))
- Train a top-down model with a detector. This is not restricted to multi-animal setups and might help to reduce the input size for the pose model. 

Let us know if these suggestions help. 

If you are still encountering problems,  could you maybe share the contents of your `pytorch_config.yaml`? This could help to further analyze potential next steps.

**Comment by deruyter92** [comment]
Closing this issue. Feel free to re-open if you are still encountering problems!
